core/src/main/java/com/amplifyframework/analytics/Properties.java

```diff
@@ -15,34 +15,152 @@
+ * different analytics services will have support for different datatypes, Properties provides a
+ * baseline of support for String, Boolean, Double, and Integer. If one of these aren't available in
+ * a given service, the plugin is expected to cast it to something supported. (e.g., convert Boolean
+ * true to the String "true")
+ *
+ * <pre>
+ *     Properties properties = Properties.builder()
+ *          .add("PostType", "UserImage")
+ *          .add("LikedUserID", 78219)
+ *          .add("FirstLike", true)
+ *          .build();
+ * </pre>
  */
-public class Properties {
+public final class Properties implements Iterable<Map.Entry<String, Property<?>>> {
+    private final Map<String, Property<?>> properties;
+
+    private Properties(Map<String, Property<?>> properties) {
+        this.properties = properties;
+    }
+
+    /**
+     * Iterator allows Properties to support foreach operations.
+     *
+     * <pre>
+     *      for (Map.Entry&lt;String, Property&lt;?&gt;&gt; entry : properties) {
+     *          String key = entry.getKey();
+     *          Property&lt;?&gt; value = entry.getValue();
+     *
+     *          // Do something with key and value
+     *      }
+     * </pre>
+     *
+     * @return An {@link Iterator} of the underlying {@link Map#entrySet()} to enable foreach access
+     */
+    @Override
+    @NonNull
+    public Iterator<Map.Entry<String, Property<?>>> iterator() {
+        return properties.entrySet().iterator();
+    }
+
+    /**
+     * size returns the number of properties.
+     *
+     * @return The number of properties
+     */
+    @NonNull
+    public Integer size() {
+        return properties.size();
+    }
 
     /**
-     * Map to hold analytics properties.
+     * get returns a specific {@link Property} by its name.
+     *
+     * @param propertyName The name of the property to retrieve
+     * @return The corresponding {@link Property}
      */
-    @SuppressWarnings("VisibilityModifier")
-    protected final Map<String, Property<?>> properties;
+    @NonNull
+    public Property<?> get(@NonNull String propertyName) {
+        return properties.get(propertyName);
+    }
 
     /**
-     * Properties constructor that initializes the underlying map.
+     * Returns a new {@link Builder} to configure an instance of Properties.
+     *
+     * @return a {@link Builder}
      */
-    public Properties() {
-        properties = new HashMap<>();
+    @NonNull
+    public static Builder builder() {
+        return new Builder();
     }
 
     /**
-     * Returns the underlying property map.
-     * @return underlying properties map
+     * Builder is used to create and configure an instance of {@link Properties}. Its
+     * methods return the Builder instance to allow for fluent method chaining.
      */
-    public final Map<String, Property<?>> get() {
-        return Immutable.of(properties);
+    public static final class Builder {
```

Comment:
Hm, I forget. Since this is part of the `core` spec: Do we need an `<T, P extends Property<T>> Builder add(String, P property)` method on this builder, as well? That way the user has an extension point, if they're using a property other than the ones our plugin does? (That is, another plugin might only work with `GizmoProperty` values.)

Reply:
Thank you for noting this. I thought about this this morning a while and wasn't sure of the best way to present an extension point. Will grab you tomorrow to chat about it.